iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VoltageLevelImpl.java

```diff
@@ -618,4 +619,91 @@ public void allocateVariantArrayElement(int[] indexes, int sourceIndex) {
+                for (Terminal oldTerminal : topologyModel.getTerminals()) {
+                    Bus connectableBus = oldTerminal.getBusBreakerView().getConnectableBus();
+                    boolean connected = oldTerminal.isConnected();
+                    oldTopologyModelInfos.add(Triple.of((TerminalExt) oldTerminal, connectableBus, connected));
+                }
+
+                for (var triple : oldTopologyModelInfos) {
+                    TerminalExt oldTerminalExt = triple.getLeft();
+                    Bus connectableBus = triple.getMiddle();
+                    boolean connected = triple.getRight();
+
+                    // if there is no way to find a connectable bus, remove the connectable
+                    // an alternative would be to connect them all to a new trash configured bus
+                    if (connectableBus != null) {
+                        AbstractConnectable<?> connectable = oldTerminalExt.getConnectable();
+
+                        // create the new terminal with new type
+                        TerminalExt newTerminalExt = null;
+                        if (oldTerminalExt.getConnectable().getType() != IdentifiableType.BUSBAR_SECTION) {
+                            newTerminalExt = new TerminalBuilder(networkRef, this, oldTerminalExt.getSide())
+                                    .setBus(connected ? connectableBus.getId() : null)
+                                    .setConnectableBus(connectableBus.getId())
+                                    .build();
+                        }
+
+                        // first attach new terminal to connectable and to new topology model
+                        if (newTerminalExt != null) {
+                            newTerminalExt.setConnectable(connectable);
+                            newTopologyModel.attach(newTerminalExt, false);
+                        }
+
+                        // then we can detach the old terminal
+                        TopologyPoint oldTopologyPoint = oldTerminalExt.getTopologyPoint();
+                        topologyModel.detach(oldTerminalExt);
+
+                        // replace the old terminal by the new terminal in the connectable
+                        connectable.replaceTerminal(oldTerminalExt, oldTopologyPoint, newTerminalExt, false);
+
+                        // also update regulating points terminal
+                        for (Referrer<Terminal> referrer : oldTerminalExt.getReferrerManager().getReferrers()) {
+                            referrer.onReferencedReplacement(oldTerminalExt, newTerminalExt);
+                        }
+                    } else {
+                        // here keep the removal notification
+                        oldTerminalExt.getConnectable().remove();
+                    }
+                }
+
+                // also here keep the notification for remaining switches and internal connection removal
```

Comment:
There are no notifications for internal connection removal
```suggestion
                // also here keep the notification for remaining switches removal
```